onnxruntime/core/optimizer/pre_shape_node_elimination.cc

```diff
@@ -0,0 +1,55 @@
+#include "core/common/logging/logging.h"
+#include "core/optimizer/rewrite_rule.h"
+#include "core/optimizer/pre_shape_node_elimination.h"
+#include "core/optimizer/utils.h"
+#include "core/graph/graph.h"
+#include "core/graph/graph_utils.h"
+
+namespace onnxruntime {
+
+Status PreShapeNodeElimination::Apply(Graph& graph, Node& node, RewriteRuleEffect& rule_effect, const logging::Logger&) const {
+  // Remove Output Edges.
+  graph_utils::RemoveNodeOutputEdges(graph, node);
+  const Node* p_input_node = graph_utils::GetInputNode(node, 0);
+
+  // Get mutable input node
+  Node& input_node = *graph.GetNode(p_input_node->Index());
+  const int output_idx = graph_utils::GetNodeOutputIndexFromOutputName(input_node, node.MutableInputDefs()[0]->Name());
+
+  // Update InputDefs of the next shape nodes.
+  auto output_nodes = graph.GetMutableConsumerNodes(node.OutputDefs()[0]->Name());
+  for (Node* next_node : output_nodes) {
+    for (auto& input_def : next_node->MutableInputDefs()) {
+      input_def = input_node.MutableOutputDefs()[output_idx];
+    }
+  }
+
+  graph.RemoveNode(node.Index());
+  rule_effect = RewriteRuleEffect::kRemovedCurrentNode;
```

Comment:
here we removed the node explicted, will the base class tries removing it again?

Reply:
The base class "RewriteRule" does not remove the node, you can remove the node by using :
graph_utils::RemoveNode(graph, node)

Or explicitly use graph.RemoveNode(node.Index());
the second one is used if you have to do extra changes on edges or input/output definitions.